src/sentry/monitors/endpoints/organization_monitor_stats.py

```diff
@@ -60,28 +53,57 @@ def get(self, request: Request, organization, project, monitor) -> Response:
         environments = get_environments(request, organization)
 
         if environments:
-            history = history.filter(monitor_environment__environment__in=environments)
-
-        for dt, status, duration in history.values_list(
-            "date_added", "status", "duration"
-        ).iterator():
-            ts = normalize_to_epoch(dt, args["rollup"])
-            stats[ts][status] += 1
-            if duration:
-                duration_stats[ts]["sum"] += duration
-                duration_stats[ts]["num_checkins"] += 1
-
-        stats_duration_data = []
-        statuses_to_name = dict(CheckInStatus.as_choices())
-        # compute average duration and construct response object
-        for ts, data in stats.items():
-            duration_sum, num_checkins = duration_stats[ts].values()
-            avg_duration = 0
-            if num_checkins > 0:
-                avg_duration = duration_sum / num_checkins
-            datapoint = {statuses_to_name[status]: data[status] for status in tracked_statuses}
-            datapoint["ts"] = ts
-            datapoint["duration"] = avg_duration
-            stats_duration_data.append(datapoint)
-
-        return Response(stats_duration_data)
+            check_ins = check_ins.filter(monitor_environment__environment__in=environments)
+
+        # Use postgres' `date_bin` to bucket rounded to our rolllups
+        bucket = Func(
+            timedelta(seconds=args["rollup"]),
+            "date_added",
+            datetime.fromtimestamp(end),
+            function="date_bin",
+            output_field=DateTimeField(),
+        )
+        # Save space on date allocation and return buckets as unix timestamps
+        bucket = Extract(bucket, "epoch")
+
+        # retrieve the list of checkins in the time range and count each by
+        # status. Bucketing is done at the postgres level for performance
+        check_in_history = (
+            check_ins.all()
+            .annotate(bucket=bucket)
+            .values("status", "bucket")
+            .order_by("bucket")
+            .annotate(count=Count("*"))
+            .values_list("bucket", "status", "count")
+        )
```

Comment:
Produces as query that looks like this

```sql
SELECT
  "sentry_monitorcheckin"."status",
  EXTRACT(
    'epoch'
    FROM
      date_bin(
        '1 day' :: interval,
        "sentry_monitorcheckin"."date_added",
        DATE('2023-06-06 09:00:00')
      ) AT TIME ZONE 'UTC'
  ) AS "bucket",
  COUNT(*) AS "count"
FROM
  "sentry_monitorcheckin"
WHERE
  (
    "sentry_monitorcheckin"."date_added" > DATE('2023-05-06 07:57:00+00:00')
    AND "sentry_monitorcheckin"."date_added" <= DATE('2023-06-06 09:57:00+00:00')
    AND "sentry_monitorcheckin"."monitor_id" = 1975
  )
GROUP BY
  "sentry_monitorcheckin"."status",
  EXTRACT(
    'epoch'
    FROM
      date_bin(
        '1 day' :: interval,
        "sentry_monitorcheckin"."date_added",
        DATE('2023-06-06 09:00:00')
      ) AT TIME ZONE 'UTC'
  )
ORDER BY
  "bucket" ASC
```

The explain on this looks like this

```


GroupAggregate (cost=1427.53..1792.54 rows=12165 width=44) (actual time=310.837..319.933 rows=67 loops=1) |  
-- | --
Group Key: (EXTRACT(epoch FROM (date_bin('1 day'::interval, date_added, ('2023-06-06'::date)::timestamp with time zone) AT TIME ZONE 'UTC'::text))), status |  
-> Sort (cost=1427.53..1457.96 rows=12171 width=36) (actual time=310.505..312.472 rows=43166 loops=1) |  
Sort Key: (EXTRACT(epoch FROM (date_bin('1 day'::interval, date_added, ('2023-06-06'::date)::timestamp with time zone) AT TIME ZONE 'UTC'::text))), status |  
Sort Method: quicksort Memory: 3560kB |  
-> Index Only Scan using sentry_moni_monitor_0a49ce_idx on sentry_monitorcheckin (cost=0.57..601.66 rows=12171 width=36) (actual time=0.176..294.836 rows=43166 loops=1) |  
Index Cond: ((monitor_id = 1975) AND (date_added > '2023-05-06'::date) AND (date_added <= '2023-06-06'::date)) |  
Heap Fetches: 39713 |  
Planning Time: 1.659 ms |  
Execution Time: 320.254 ms
```

(This was done on an arbitrary monitor, note the month range)